drift/lib/src/runtime/manager/references.dart

```diff
@@ -0,0 +1,224 @@
+part of 'manager.dart';
+
+/// A simple function for generating aliases for referenced columns
+///
+/// This function is used internally by generated code and should not be used directly.
+// ignore: non_constant_identifier_names
+String $_aliasNameGenerator(
+    GeneratedColumn currentColumn, GeneratedColumn referencedColumn) {
+  return '${currentColumn.tableName}__${currentColumn.name}__${referencedColumn.tableName}__${referencedColumn.name}';
+}
+
+/// Base class for the "WithReference" classes
+///
+/// When a user calls `withReferences` on a manager, the item is returned, along with an instance of this
+/// class which contains getters for the referenced tables manager which are pre-filtered to the item.
+/// So the following:
+/// ```dart
+/// final (group,refs) = await groups.filter((f) => f.id(5)).withReferences().getSingle()
+/// final usersInGroup = refs.users.get() // filter((f) => f.group(group.id)) is already applied
+/// ```
+/// is short for:
+/// ```dart
+/// final group = await groups.filter((f) => f.id(5)).getSingle()
+/// final usersInGroup = await users.filter((f) => f.group(group.id)).get()
+/// ```
+/// {@macro manager_internal_use_only}
+class BaseReferences<$Database extends GeneratedDatabase, $Table extends Table,
+    $Dataclass> {
+  /// The database instance
+  // ignore: non_constant_identifier_names
+  final $Database $_db;
+
+  /// The table of this item
+  // ignore: non_constant_identifier_names
+  final TableInfo<$Table, $Dataclass> $_table;
+
+  /// The raw [TypedResult] for this item
+  // ignore: non_constant_identifier_names
+  final TypedResult $_typedResult;
+
+  // The item these references are for
+  // ignore: public_member_api_docs, non_constant_identifier_names
+  late final $Dataclass $_item = $_typedResult.readTable($_table);
+
+  /// Create a [BaseReferences] class
+  // ignore: non_constant_identifier_names
+  BaseReferences(this.$_db, this.$_table, this.$_typedResult);
+}
+
+/// Type definition for a function that transforms the state of a manager
+typedef StateTransformer = T Function<
+    T extends TableManagerState<dynamic, dynamic, dynamic, dynamic, dynamic,
+        dynamic, dynamic, dynamic, dynamic, dynamic>>(T $state);
+
+T _defaultStateTransformer<
+    T extends TableManagerState<dynamic, dynamic, dynamic, dynamic, dynamic,
+        dynamic, dynamic, dynamic, dynamic, dynamic>>(T $state) {
+  return $state;
+}
```

Comment:
It might be too late for that, and in this case I think it's fine, but since managers are mostly internal and short-lived builders, I think it would be much easier to just make them mutable.

We need all this ceremony with the type parameters just to indicate that this function turns specific manager instances into the something of the same type - with mutable managers it could just be `void Function(TableManagerState)`.

Reply:
I agree, but IMHO it's a worthy tradeoff.

There are some places where it would be simplified, but this manager api is quite complex, and I don't trust myself enough not to have bugs.